### PR TITLE
Community seeds

### PR DESCRIPTION
**Pull request by MattEqualsCoder**
Adding the ability to have more granular control over what gets generated where and adding the ability to share configs, opening the possibility for community created seeds for streamers to play.

- Every location can now have either progression items, junk items, or specific items be specified
- All progression items can be specified as early items
- Updated rom generation to handle potentially unwinnable seeds gracefully by displaying an error message upon generating the rom
  - Depending on settings, sometimes you can try regenerating the seed when this error displays, and it'll work successfully. If you requested a particular item at a location where the item isn't needed to get there, it may have randomly picked a logic chain that is impossible.
- Base 64 encoded settings strings can be generated and shared that house all of the config data
  - This can be retrieved by either right clicking in the rom list page or by viewing the spoiler log
  - All customization such as sprites, cas'troid, etc. should pull from the local player's settings. Only seed info is copied.

### CODE REVIEW

**Comment by Vivelin** [comment]
I can't really give a thorough review right but I didn't see anything too out of the ordinary so feel free to continue

**Comment by MattEqualsCoder** [comment]
@Vivelin After pulling my hair out for hours, I think I've got this pretty stable. There were a mountain of corner cases that were arising when I was trying to run stats or unit tests that revolving around how it shuffled items around. Obviously still certain combinations people put in may be unbeatable, but it should catch those situations upon generation and warn people.

I'm going to do some more testing tomorrow, but otherwise this should be good to go.

**Comment by Vivelin** [comment]
I'll give it a review in a minute. It might be best to release this soon-ish (mark it as pre-release for now?) so there's a little more time to test. I've had some other small changes I wanted to make but we've already got more than enough to warrant a new release, I think.

File: src/Randomizer.App/Windows/GenerateRomWindow.xaml [hunk 1]
```diff
@@ -271,6 +190,11 @@
                 <TextBox x:Name="SeedInput"
                          Text="{Binding Seed}" />
               </controls:LabeledControl>
+
+              <controls:LabeledControl Text="Config string (optional):">
```
Comment (Vivelin) [inline src/Randomizer.App/Windows/GenerateRomWindow.xaml:194]: "Config string" might not be the best name for this in the UI... Maybe import/export buttons? Or "Copy settings" and "Import from clipboard" or something like that?

File: src/Randomizer.App/Windows/GenerateRomWindow.xaml.cs [hunk 2]
```diff
@@ -352,5 +501,81 @@ private void LogicCheckBox_Checked(object sender, RoutedEventArgs e)
...
+        /// </summary>
+        /// <param name="sender">The dropdown that was updated</param>
+        /// <param name="e">The event object</param>
+        private void LocationsRegionFilter_SelectionChanged(object sender, SelectionChangedEventArgs e)
+        {
+            var comboBox = sender as ComboBox;
+            var selectedRegion = comboBox.SelectedItem as string;
+            foreach (FrameworkElement obj in LocationsGrid.Children) {
+                var location = obj.Tag as Location;
+                obj.Visibility = selectedRegion.Contains(location.Region.Name) ? Visibility.Visible : Visibility.Collapsed;
```
Comment (Vivelin) [inline src/Randomizer.App/Windows/GenerateRomWindow.xaml.cs:516]: This can cause a null reference exception if Tag is ever not a location or null.

**Review by Vivelin — DISMISSED** [review]

**Comment by MattEqualsCoder** [comment]
Fixed all of the above and added the logic config to the spoiler log

**Review by Vivelin — APPROVED** [review]
